code/game/mecha/mecha_control_console.dm

```diff
@@ -14,62 +14,60 @@
 	if(!ui)
-		ui = new(user, src, ui_key, "exosuit_control.tmpl", "Exosuit Control Console", 420, 500)
+		ui = new(user, src, ui_key, "MechaControlConsole", name, 420, 500, master_ui, state)
 		ui.open()
-		ui.set_auto_update(1)
-
-/obj/machinery/computer/mecha/ui_data(mob/user, ui_key = "main", datum/topic_state/state = GLOB.default_state)
-	var/data[0]
-	data["screen"] = screen
-	if(screen == 0)
-		var/list/mechas[0]
-		var/list/trackerlist = list()
-		for(var/stompy in GLOB.mechas_list)
-			var/obj/mecha/MC = stompy
-			trackerlist += MC.trackers
-		for(var/thing in trackerlist)
-			var/obj/item/mecha_parts/mecha_tracking/TR = thing
-			var/answer = TR.get_mecha_info()
-			if(answer)
-				mechas[++mechas.len] = answer
-		data["mechas"] = mechas
-	if(screen == 1)
-		data["log"] = stored_data
-	return data
-
-/obj/machinery/computer/mecha/Topic(href, href_list)
-	if(..())
-		return 1
-
-	var/datum/topic_input/afilter = new /datum/topic_input(href,href_list)
-	if(href_list["send_message"])
-		var/obj/item/mecha_parts/mecha_tracking/MT = afilter.getObj("send_message")
-		var/message = strip_html_simple(input(usr,"Input message","Transmit message") as text)
-		if(!trim(message) || ..())
-			return 1
-		var/obj/mecha/M = MT.in_mecha()
-		if(M)
-			M.occupant_message(message)
 
-	if(href_list["shock"])
-		var/obj/item/mecha_parts/mecha_tracking/MT = afilter.getObj("shock")
-		MT.shock()
+/obj/machinery/computer/mecha/tgui_data(mob/user)
+	var/list/data = list()
+	data["beacons"] = list()
+	var/list/trackerlist = list()
+	for(var/stompy in GLOB.mechas_list)
+		var/obj/mecha/MC = stompy
+		trackerlist += MC.trackers
+	for(var/thing in trackerlist)
+		var/obj/item/mecha_parts/mecha_tracking/TR = thing
+		var/list/tr_data = TR.tgui_data(user)
+		if(tr_data)
+			data["beacons"] += list(tr_data)
+
+	LAZYINITLIST(stored_data)
+	data["stored_data"] = stored_data
 
-	if(href_list["get_log"])
-		var/obj/item/mecha_parts/mecha_tracking/MT = afilter.getObj("get_log")
-		stored_data = MT.get_mecha_log()
-		screen = 1
+	return data
 
-	if(href_list["return"])
-		screen = 0
 
-	SSnanoui.update_uis(src)
-	return
+/obj/machinery/computer/mecha/tgui_act(action, params)
+	if(..())
+		return
+	switch(action)
+		if("send_message")
+			var/obj/item/mecha_parts/mecha_tracking/MT = locateUID(params["mt"])
+			if(istype(MT))
+				var/message = strip_html_simple(input(usr, "Input message", "Transmit message") as text)
+				if(!message || !trim(message) || ..())
+					return FALSE
+				var/obj/mecha/M = MT.in_mecha()
+				if(M)
+					M.occupant_message(message)
+				return TRUE
+		if("shock")
+			var/obj/item/mecha_parts/mecha_tracking/MT = locateUID(params["mt"])
+			if(istype(MT))
+				MT.shock()
+				return TRUE
+		if("get_log")
+			var/obj/item/mecha_parts/mecha_tracking/MT = locateUID(params["mt"])
+			if(istype(MT))
+				stored_data = MT.get_mecha_log()
+				return TRUE
+		if("clear_log")
+			stored_data = null
```

Comment:
I suggest doing `stored_data = list()` instead of setting it to null. This will allow you to avoid calling `LAZYINITLIST(stored_data)` on line 38.

Also on line 11: `var/stored_data` should be changed to `var/list/stored_data = list()`